FIP-TBD:  More API between user-programmed actors and built-in miner actor for building decentralized staking protocol

jennijuju opened the pull request:
Export additional built-in miner actor methods for invocation by user actors, so to enable more decentralized DeFi protocol for Storage Provider (SP) services.

Stebalien commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+- GetAvailableBalance
+- GetLockedReward
+- GetLockedInitialPledge
+- GetExpectedRewards
```
This depends on

1. `ThisEpochReward` from the reward actor.
2. Some method on the power actor to get the QA power for a miner.

luckyparadise commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+---
+fip: "tbd"
```
@jennijuju could you please name this as FIPXXXX as is the standard practice? Thank you.

jennijuju replied:
Meh I don’t think this one in particular matters - given this won’t ever be merged

luckyparadise commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+author:  @Schwartz10, @jennijuju
+discussions-to: 
+status: Draft
+type: https://github.com/filecoin-project/FIPs/discussions/1034
```
The type here should be "Technical" except otherwise. And please move this discussion link to the correct placeholder under "discussions-to".

luckyparadise commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+---
+
+
+# FIP-TBD:  More API between user-programmed actors and built-in miner actors for building decentralized staking protocol
```
This should be FIPXXXX followed by the title.

jennijuju replied:
will not take this suggestion as I expect a fip number here if it will be merged.

luckyparadise commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+
+## Product and Incentive Considerations
+
+This FIP enables more trustless and decentralized staking and DeFi protocol for token holders to partipate in. Staking and Defi protocol can drive Filecoin onchain activities, improve FIL utilities and enabling more (1) storage providers to get liquidity to join the network and prvoide storage services & secure the network consensus (2) token holders to put their FIL into use and improve Filecoin ...
```
> to get liquidity to join the network and **prvoide** storage services & secure the network consensus. 

Nit spelling correction for "provide".

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+---
+fip: "tbd"
+title: More API between user-programmed actors and built-in miner actor for building decentralized staking protocol 
```
```suggestion
title: More APIs between user-programmed actors and built-in miner actor for building decentralized staking protocol 
```
nit

anorth replied:
The real key here is exposing sector termination, which is buried in this title. I suggest something like "Export sector termination method from miner actor", and it's ok if a couple of little additions come in with it (though they could also be a separate FIP)

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+---
+
+
+# FIP-TBD:  More API between user-programmed actors and built-in miner actors for building decentralized staking protocol
```
```suggestion
# FIPXXXX:  More APIs between user-programmed actors and built-in miner actors for building decentralized staking protocol
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+
+## Simple Summary
+
+Export additional built-in miner actor methods for invocation by user actors, so to enable more decentralized DeFi protocol for Storage Provider (SP) services.
```
```suggestion
Export additional built-in miner actor methods for invocation by user actors to enable more decentralized DeFi protocols for Storage Provider (SP) services.
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+
+`Vec<TerminationDeclaration>` can be nullable. When it is null, flush/terminate the sectors that are queued up in the cron for termination.
+
+To prevent the same sectors to be added to the termination queue in the cron and consume block space maliciously, calling this method will immediately terminate the sectors in early_terminations queue first, then execute termination of  the sectors submitted in TerminationDeclaration. 
```
> early_terminations queue

Is there a resource about this we can link to?  (I assume not given https://spec.filecoin.io/ being outdate? )

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+
+Following the FIP-0050 specification for exported builin actor methods in, we add a set of additional methods.
+
+### Operation 
+
+**TerminateSectors** In Miner Actor
```
```suggestion
### **TerminateSectors** In Miner Actor
```

Idea of just making this the heading for this section

BigLep commented on FIPS/fip-tbd.md:
````diff
@@ -0,0 +1,106 @@
+```
+
+Update the return value of the method to include sector numbers that are successfully terminated or still pending in the queue.. 
+
````
Do we need discussion about how TerminateSector will be called by a 3rd party?  Does the smart contract have some derivation of the SPs signing key.  Sorry if I'm using the wrong terms, and maybe that is obvious to everyone, so feel free to disregard.

anorth replied:
It definitely needs to be restricted to the miner's control addresses

jennijuju replied:
In the beginning of this session, I mentioned that we are following the existing terminate sector behaviour unless proposed otherwise - and the caller to be restricted to the miners control address is an existing behaviour. Does fip editors prefer me to explicitly state that?

anorth replied:
I think there's some confusion about whether this is exporting the existing TerminateSectors method, or a new one, or modifying the existing one (for all callers) and then exporting it. It would help to spell this out very explicitly.

If it's the existing method, then yes this is existing behaviour. My opinion is we should make a new method that's much simpler, in which case we should specify things like allowed callers.

jennijuju replied:
Ack - will do!

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+
+This FIP introduces new builtin actor methods, therefore needs a new actors version shipped in a network upgrade. No breaking changes to existing methods, however.
+
+## Test Cases
```
Should we have a TODO to enumerate this out more?  I assume we ultimately want a lit of asserts to make in the test cases to make sure everyone is aligned.

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,106 @@
+
+## Security Considerations
+
+No new method is being introduced that can be used to attack the network, and bugs in the implementation will only lead to incorrect behavior being observed by user actors.
```
Are we sure?  Isn't having a smart contract being able to invoke terminatesector on an SP's behalf a big deal.

jennijuju replied:
added to the secuirty consideration

anorth replied:
This line doesn't seem right to me at all. The method could be used to attack the network if it has a problem, and bugs could lead to all kinds of unexpected behaviour. But nevertheless, that's not the kind of thing we've bothered writing in any FIP that introduces a new method in the past. This isn't any safer or more dangerous than other FIPs. I don't think this sentence belongs at all.

jennijuju replied:
Will remove!

BigLep reviewed (commented):
Likely a bunch of newbie comments.  Don't feel like you need to explain things to me if they aren't broadly applicable.  I can learn / ask offline..

Also, if it's of use, I'm happy to take a cohesive pass on some of the grammar.  I started but realized that may only be worthwhile if this FIP looks to be on track to get accepted.

jennijuju commented:
> Also, if it's of use, I'm happy to take a cohesive pass on some of the grammar. I started but realized that may only be worthwhile if this FIP looks to be on track to get accepted.

I would appreciate a grammar review when this fip is closer to “merge draft” stage! I will do a grammar pass myself as well, wanted to get a quick gut check on the ideas first!

jennijuju commented:
Have a redesign on the spec - will reopen once updated!

jennijuju commented:
Updated the FIP with the peer feedback received and addressed some editor feedback! I believe this FIP is ready for another round of Peer and editor review now!

cc @anorth @Stebalien  @filecoin-project/fips-editors @BigLep

kaitlin-beegle commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,112 @@
+
+## Abstract
+
+Upon the FEVM launch, [FIP-0050](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0050.md) was also introduced with a list of the builtin actor methods that are exposed as callables from user actors. Since then,  we’ve seen a number of protocols build lending or leasing applications around Miner Actors collateral & rewards. As common practice in DeFi, protocols often introduce a “liq ...
```
```suggestion
Following FEVM launch, [FIP-0050](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0050.md) introduced a list of builtin actor methods that are exposed as callables from user actors. Since then, we have seen a number of protocols use miner actor collateral and/or rewards to build lending and leasing applications. 

It is a common practice in DeFi applications to manage risk for liquidity providers (i.e., those lending money) by "liquidating" collateral supplied by borrowers. In the Filecoin ecosystem, 'liquidation' is achieved when Storage Providers terminate a miner's active sectors, pay the required penalty, recover their pledged collateral, and repay liquidity providers. This process is a critical part of FEVM economic security, as is understanding the precise value of existing collateral. 

```
Updated language for clarity!

kaitlin-beegle commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,112 @@
+
+Upon the FEVM launch, [FIP-0050](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0050.md) was also introduced with a list of the builtin actor methods that are exposed as callables from user actors. Since then,  we’ve seen a number of protocols build lending or leasing applications around Miner Actors collateral & rewards. As common practice in DeFi, protocols often introduce a “liq ...
+
+This FIP aims to introduce an exported sector termination method in miner actor for supporting onchain liquidation procedures for decentralized staking protocol. 
```
```suggestion
This FIP aims to introduce an exported sector termination method in miner actor to support on-chain liquidation procedures. 
```

kaitlin-beegle commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,112 @@
+
+There are a few motivations to this change:
+
+- Trustless DeFi - it’s important that users of a DeFi protocol can trust code and not humans to protect their assets. When humans are in the loop, there is necessarily a “judgment call zone”, which violates all aspects of DeFi, it also creates personal safety issues for individuals who are known to process liquidations with no plausible deniability. 
```
I am not sure how this statement supports the changes proposed by this FIP.  Is the correct interpretation that exporting these actors simply allows for more/better DeFi operations to be executed via code?

kaitlin-beegle commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,112 @@
+
+To allow caller (storage providers, stake pools and etc) to have more precise control over the amount of the sectors to be terminated based on their operational needs, this FIP introduces a new `TerminateSector2` method. This method follows most of the behaviour of the existing `TerminateSectors` method (method 9), notably:
+- Only a control address of the miner actor can call this method.
+- Attempt to terminate sectors that is in the current or the next proving deadline will fail with `USR_ILLEGAL_ARGUMENT`.
```
```suggestion
- Attempt to terminate sectors that are in the current or the next proving deadline will fail with `USR_ILLEGAL_ARGUMENT`.
```

kaitlin-beegle reviewed (commented):
Added review for grammar and clarity.

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+
+There are a few motivations to this change:
+
+- Trustless DeFi - it’s important that users of a DeFi protocol can trust code and not humans to protect their assets. When humans are in the loop, there is necessarily a “judgment call zone”, which violates all aspects of DeFi, it also creates personal safety issues for individuals who are known to process liquidations with no plausible deniability. 
```
```suggestion
- Trustless DeFi - it’s important that users of a DeFi protocol can trust code and not humans to protect their assets. When humans are in the loop, there is necessarily a “judgment call zone”, which violates all aspects of DeFi.  It also creates personal safety issues for individuals who are known to process liquidations with no plausible deniability. 
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+- Liability concern - from the DeFi protocol creator’s perspective, being an external third party that has to process liquidations off chain (rather than creating a keeper network of liquidators) introduces liability and regulatory concern to the protocol, making it harder to operate in certain jurisdictions 
+- Incentive design - by designing their own rulesets for liquidations, application developers can incentivize keepers to protect their protocol in a decentralized manner
+
+This FIPs propose a builtin actors APIs that can trigger sector termination in smart contracts that staking protocols often need to build offchain logics/oracles for currently. 
```
```suggestion
This FIP proposes a builtin actors APIs that can trigger sector termination in smart contracts that staking protocols often need to build offchain logics/oracles for currently. 
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+
+**TerminateSectors2** In Miner Actor
+
+To allow caller (storage providers, stake pools and etc) to have more precise control over the amount of the sectors to be terminated based on their operational needs, this FIP introduces a new `TerminateSector2` method. This method follows most of the behaviour of the existing `TerminateSectors` method (method 9), notably:
```
```suggestion
To allow callers (e.g., storage providers, stake pools) to have more precise control over the amount of the sectors to be terminated based on their operational needs, this FIP introduces a new `TerminateSector2` method. This method follows most of the behaviour of the existing `TerminateSectors` method (method 9), notably:
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+
+To allow caller (storage providers, stake pools and etc) to have more precise control over the amount of the sectors to be terminated based on their operational needs, this FIP introduces a new `TerminateSector2` method. This method follows most of the behaviour of the existing `TerminateSectors` method (method 9), notably:
+- Only a control address of the miner actor can call this method.
+- Attempt to terminate sectors that are in the current or the next proving deadline will fail with `USR_ILLEGAL_ARGUMENT`.
```
```suggestion
- Attempts to terminate sectors that are in the current or the next proving deadline will fail with `USR_ILLEGAL_ARGUMENT`.
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+- The method will always process the sector(s) that are already in the `early_termination` queue before processing newly submitted sectors.
+
+This FIP proposes the following behaviour changes:
+- Currently, `TerminateSectors` allows callers to submit early termination for at most 3000 sectors in one message. `TerminateSectors2` will remove this hardcoded upper bound. Instead, `TerminateSectors2` takes a `max_termination` parameter for caller to specify the maximum number of sectors to be terminated. 
```
```suggestion
- Currently, `TerminateSectors` allows callers to submit early termination for at most 3000 sectors in one message. `TerminateSectors2` will remove this hardcoded upper bound. Instead, `TerminateSectors2` takes a `max_termination` parameter for callers to specify the maximum number of sectors to be terminated. 
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+- Currently, `TerminateSectors` allows callers to submit early termination for at most 3000 sectors in one message. `TerminateSectors2` will remove this hardcoded upper bound. Instead, `TerminateSectors2` takes a `max_termination` parameter for caller to specify the maximum number of sectors to be terminated. 
+- `TerminateSectors2` will always first terminate the sectors that are already in the `early_termination` queue (either added through cron or through `TerminateSectors(2)` messages), then process the new sectors submitted in the message. This prevents the same sectors to be added to the termination queue in the cron and consume block space maliciously.
+ - A newly submitted sector will be skipped if it was already in the `early_termination` queue and got executed first, so that the sector will not be terminated twice. 
+- `TerminateSectors2` will only terminate sectors up to the number caller specified in the `max_termination` parameter. If there are more sectors in `early_termination` queue and the new sectors submitted in the message than the `max_termination` amount, the message will fail with `USR_ILLEGAL_ARGUMENT`. 
```
```suggestion
- `TerminateSectors2` will only terminate sectors up to the number specified in the `max_termination` parameter. If there are more sectors in `early_termination` queue and the new sectors submitted in the message than the `max_termination` amount, the message will fail with `USR_ILLEGAL_ARGUMENT`. 
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+- `TerminateSectors2` will only terminate sectors up to the number caller specified in the `max_termination` parameter. If there are more sectors in `early_termination` queue and the new sectors submitted in the message than the `max_termination` amount, the message will fail with `USR_ILLEGAL_ARGUMENT`. 
+  - Caller can query the `early_termination` queue size ahead of submitting the message to determine the maximum number of sectors and/or new sectors it can terminate in one message.
+  - That said, `TerminateSectors2` will not add sectors to the termination queue.
+- `TerminateSectors2` will fail with `SYS_OUT_OF_GAS` and abort the all execution if the caller doesn't have enough gas to execute the termination and/or the message itself run out of gas.
```
```suggestion
- `TerminateSectors2` will fail with `SYS_OUT_OF_GAS` and abort all execution if the caller doesn't have enough gas to execute the termination and/or the message itself runs out of gas.
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+
+## Backwards Compatibility
+
+This FIP introduces new builtin actor methods, therefore needs a new actors version shipped in a network upgrade. No breaking changes to existing methods, however, existing methods may be deprecated in future network upgrades.
```
```suggestion
This FIP introduces new builtin actor methods, therefore needs a new actors version shipped in a network upgrade. There are no breaking changes to existing methods, however, existing methods may be deprecated in future network upgrades.
```

BigLep commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+
+To prevents the same sectors to be added to the termination queue in the cron and consume block space maliciously with programmatic smart contracts, `TerminateSectors2` will not add any sectors to the cron queue and it will always terminate the sectors that are already in the `early_termination` queue first, then process the new sectors submitted in the message and skip the sectors that are alrea ...
+
+While the newly introduced `TerminateSectors2` method allows a smart contract to terminate a sector on SPs behalf, it still requires the SP to add the smart contract as a control address first. When SPs signs and send the message to add smart contract as a control address, this should be considered as a high trust action, as the SP is authorizing the control right to the smart contract. 
```
```suggestion
While the newly introduced `TerminateSectors2` method allows a smart contract to terminate a sector on an SP's behalf, it still requires the SP to add the smart contract as a control address first. When a SP signs and sends the message to add smart contract as a control address, this should be considered as a high trust action, as the SP is authorizing the control right to the smart contract. 
```

BigLep reviewed (commented):
I read for my own learning, and in the process made a few gramatical suggestions.

anorth commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+- Currently, `TerminateSectors` allows callers to submit early termination for at most 3000 sectors in one message. `TerminateSectors2` will remove this hardcoded upper bound. Instead, `TerminateSectors2` takes a `max_termination` parameter for caller to specify the maximum number of sectors to be terminated. 
+- `TerminateSectors2` will always first terminate the sectors that are already in the `early_termination` queue (either added through cron or through `TerminateSectors(2)` messages), then process the new sectors submitted in the message. This prevents the same sectors to be added to the termination queue in the cron and consume block space maliciously.
+ - A newly submitted sector will be skipped if it was already in the `early_termination` queue and got executed first, so that the sector will not be terminated twice. 
+- `TerminateSectors2` will only terminate sectors up to the number caller specified in the `max_termination` parameter. If there are more sectors in `early_termination` queue and the new sectors submitted in the message than the `max_termination` amount, the message will fail with `USR_ILLEGAL_ARGUMENT`. 
```
I know we discussed this a lot but I have forgotten the various reasoning. Why does this fail if there were no new sectors submitted for termination but the queue is longer than max_terminations? I think we should allow manually terminating part of the queue even if it's too long to ever do in a single message.

Failing if it can't terminate the sectors that were submitted extra (because the queue is too long) is ok

jennijuju replied:
> I think we should allow manually terminating part of the queue even if it's too long to ever do in a single message.

this makes sense.

anorth commented on FIPS/fip-tbd.md:
````diff
@@ -0,0 +1,115 @@
+
+```rust
+pub struct TerminateSectors2Params {
+ pub terminations: Option<Vec<TerminationDeclaration>>,
````
Do we need both optional and empty vector? How does the behaviour of an option with `Some([])` differ from submitting `None`? Would just a vector suffice?

anorth commented on FIPS/fip-tbd.md:
```diff
@@ -0,0 +1,115 @@
+}
+
+pub struct TerminateSectors2Return {
+    // Set to true if all early termination work and new sectors termination work have been completed. Set to false otherwise.
```
This needs more explanation of how this return value could be false and when the entire method call fails instead. Again, I know we discussed this a while ago, but I have forgotten the outcomes.

anorth reviewed (commented):
The Design Rationale section is missing. It needs to describe why a new method, and the use cases motivating the all-or-nothing behaviour (that I know we discussed, but I have forgotten).

jennijuju commented:
Unfortunately, I won't be able to work on this FIP for another couple of weeks. If anyone from the community wants to add to it or take it over, or propose another FIP for this, feel free to! In the meantime, I will covert this PR to a draft and reopen it when i get to pick it up again.

Tanisha-fil reviewed (changes requested):
Given that the author and champion of this FIP doesn't currently have the bandwidth to move it forward, I suggest we mark this proposal as Stagnant in line with the official FIP process.

This acknowledges the proposal is inactive for now but can be revived in the future if a new champion emerges.

rvagg commented:
I'm fine with that but it would need a number to merge this and not just perpetually leave it in draft or just close it.
